Centralized capabilities service

hypeJunction opened the pull request:
**feature(entities): adds user capabilities service and tests**
Adds `\Elgg\UserCapabilities` service and proxies `ElggEntity::can*()` methods.
Adds PHPUnit tests for user capabilities
Creates an easier way to mock entities during testing with `\Elgg\Tests\EntityMocks`
class, which can be instantiated with the instance of PHPUnit test case.

hypeJunction committed:
feature(entities): adds user capabilities service and tests

Adds \Elgg\UserCapabilities service and proxies ElggEntity::can*() methods.
Adds PHPUnit tests for user capabilities

Create an easier way to mock entities during testing with \Elgg\Tests\EntityMocks
class, which can be instantiated with the instance of PHPUnit test case.

hypeJunction commented on engine/classes/Elgg/UserCapabilities.php:
```diff
@@ -0,0 +1,301 @@
+	 * @see elgg_set_ignore_access()
+	 */
+	public function canEditMetadata(ElggEntity $entity, $user_guid = 0, ElggMetadata $metadata = null) {
+		if (!$entity->guid) {
```
Let's revisit. We should perhaps also check if metadata belongs to this entity in the first place.

hypeJunction replied:
Will open an issue.

hypeJunction commented on engine/lib/entities.php:
```diff
@@ -271,43 +271,19 @@ function update_subtype($type, $subtype, $class = '') {
-	try {
-		$user = _elgg_services()->entityTable->getUserForPermissionsCheck($user_guid);
-	} catch (UserFetchFailureException $e) {
+	if (!$container) {
```
I am not sure how this hasn't xaused problems before.  No page owner would pass empty container to the hook.

hypeJunction commented on engine/classes/Elgg/UserCapabilities.php:
```diff
@@ -0,0 +1,301 @@
+	 */
+	public function canAnnotate(ElggEntity $entity, $user_guid = 0, $annotation_name = '') {
+		if (!empty($annotation_name) && !is_string($annotation_name)) {
+			throw new \InvalidArgumentException(__METHOD__ . ' expects \$annotation_name to be a string');
```
This is new. I think it makes sense here.

mrclay replied:
This would still accept an array. Do we really need to accept `null` and `false`?

mrclay replied:
Rather be explicit about it:
```php
if ($annotation_name === null || $annotation_name === false) {
    // accepting these for BC
    $annotation_name = '';
} elseif (!is_string($annotation_name)) {
    throw ...
}
```

mrclay commented on engine/classes/Elgg/UserCapabilities.php:
```diff
@@ -0,0 +1,301 @@
+ * @access private
+ * @since 2.2
+ */
+class UserCapabilities {
```
Did moving these methods make testing a lot easier?

hypeJunction replied:
It did

mrclay commented on engine/tests/phpunit/Elgg/UserCapabilitiesTest.php:
```diff
@@ -0,0 +1,512 @@
+		$this->assertFalse($user->canComment(33));
+		$this->assertTrue($object->canComment(33));
+		$this->assertFalse($group->canComment(33));
+		$this->assertNull($entity->canComment(33));
```
Why not false?

hypeJunction replied:
I don't know. That's what it is design. The value is then checked by extending classes. I am planning to tackle this wihen I work on commenting permissions.

mrclay commented on engine/tests/phpunit/Elgg/UserCapabilitiesTest.php:
```diff
@@ -0,0 +1,512 @@
+			'owner_guid' => 41,
+		]);
+		$this->setupServices();
+		$entity->canAnnotate(41); // owner
```
Missing assertions?

mrclay commented:
I like a lot that this centralizes the capabilities logic.

The complexity of the test mocking makes me nervous, but since the capabilities mostly depends on hooks and attributes (not deep queries) I think this is a good addition. Would be nice to remove duplicate test functionality from simpletests. I'll try to identify some of that.

mrclay commented on engine/tests/phpunit/Elgg/UserCapabilitiesTest.php:
```diff
@@ -0,0 +1,512 @@
+	/**
+	 * Setup a mock entity
+	 */
+	protected function setupEntity($guid, $type, $subtype, array $attributes = []) {
```
Can later be moved to an entityMocks service

hypeJunction replied:
Yes. This was the biggest challenge, but we now have a foundation to work with

mrclay commented:
This should ideally replace:

- [ ] ElggAnnotationTest
- [ ] \ElggCoreEntityTest::testCanAnnotateDefault
- [ ] \ElggCoreEntityTest::testCanAnnotateCallsSpecificThenGenericHook
- [ ] \ElggCoreEntityTest::testCanAnnotateHookParams
- [ ] \ElggCoreEntityTest::testCanAnnotateDoesntCallSpecificThenGenericHookForEmptyString
- [ ] \ElggCoreRegressionBugsTest::test_can_write_to_container

hypeJunction commented:
@mrclay 
Simplified mocked by adding an extra class 
Added `canEditAnnotation` and removed redundant annotation tests. 
Updated some uses of `can_write_to_container()` (not really sure what the pages plugin is doing there, so please take a look)
Not touching `\ElggCoreRegressionBugsTest::test_can_write_to_container` as it looks like it's relying on group hooks for write access.

hypeJunction commented:
Pages plugin is another use case for `container_logic_permissions`.

FYI, I am headed to the airport, and won't be around much this week, so feel free to hijack any open pull requests and do your magic. Bis gleich.

hypeJunction commented on engine/tests/phpunit/Elgg/UserCapabilitiesTest.php:
```diff
@@ -0,0 +1,694 @@
+	 * @group UserCapabilities
+	 * @\expectedException UserFetchFailureException
+	 */
+	public function testThrowsExceptionForCanObjectEditSelf() {
```
@mrclay Could you peep this. I am not sure why the class is not autoloaded.

hypeJunction commented on mod/blog/actions/blog/save.php:
```diff
@@ -91,8 +91,8 @@
 			// this can't be empty or saving the base entity fails
 			if (!empty($value)) {
-				if (can_write_to_container($user->getGUID(), $value)) {
-					$values[$name] = $value;
```
Oops

mrclay commented:
Would this cover #8790?

hypeJunction commented:
I will need to look after a good night sleep, but my gut says logic check would cover the case. Can try approaching with tests to see what happens.

hypeJunction commented:
@mrclay I think #8790 is better off with the solution you proposed. Using this would still be misusing container permissions.

hypeJunction commented:
@mrclay I have removed extra features from here. Let's get this in and then I will make new PRs for the rest of the stuff. It was getting hard to work with multiple commits.

mrclay commented:
Porting of permissions tests always makes me nervous, so I'm going to split this into 2 PRs, first that adds API and passes old tests, 2nd that ports tests.

mrclay commented:
Replaced by #9736 and  #9737

mrclay commented:
...which are both in!

hypeJunction commented:
Awesome. `container_logic_check` will follow.
